File: app/src/lib/cifrado.js
// app/src/lib/cifrado.js
// Cifrado de la sincronización: AES-256-GCM (WebCrypto). Sobre = IV (12 bytes) + texto cifrado
// con etiqueta. El contenido lleva delante la hora (8 bytes, ms, big-endian): se rechaza lo que
// tenga más de EDAD_MAX de diferencia (contra repeticiones). Mismo formato que receptor/sincro.
const AAD = new TextEncoder().encode('canvas-sincro-v1')
export const EDAD_MAX = 5 * 60 * 1000

export const b64 = {
  a(u) { let s = ''; for (let i = 0; i < u.length; i += 0x8000) s += String.fromCharCode(...u.subarray(i, i + 0x8000)); return btoa(s) },
  de(s) { return Uint8Array.from(atob(s.trim()), c => c.charCodeAt(0)) }
}

export const claveNueva = () => b64.a(crypto.getRandomValues(new Uint8Array(32)))

export async function importarClave(claveB64) {
  const raw = b64.de(claveB64)
  if (raw.length !== 32) throw new Error('La clave debe tener 32 bytes')
  return crypto.subtle.importKey('raw', raw, 'AES-GCM', false, ['encrypt', 'decrypt'])
}

export async function cifrarBytes(clave, bytes, iv = crypto.getRandomValues(new Uint8Array(12)), ahora = Date.now()) {
  const plano = new Uint8Array(8 + bytes.length)
  new DataView(plano.buffer).setBigUint64(0, BigInt(ahora))
  plano.set(bytes, 8)
  const ct = new Uint8Array(await crypto.subtle.encrypt({ name: 'AES-GCM', iv, additionalData: AAD }, clave, plano))
  const out = new Uint8Array(12 + ct.length)
  out.set(iv)
  out.set(ct, 12)
  return out
}

export async function descifrarBytes(clave, sobre, ahora = Date.now()) {
  if (sobre.length < 12 + 8 + 16) throw new Error('Clave incorrecta o mensaje alterado')
  let plano
  try {
    plano = new Uint8Array(await crypto.subtle.decrypt({ name: 'AES-GCM', iv: sobre.subarray(0, 12), additionalData: AAD }, clave, sobre.subarray(12)))
  } catch { throw new Error('Clave incorrecta o mensaje alterado') }
  const t = Number(new DataView(plano.buffer, plano.byteOffset).getBigUint64(0))
  if (Math.abs(ahora - t) > EDAD_MAX) throw new Error('Mensaje vencido (revisa la hora del celular y de la PC)')
  return plano.subarray(8)
}

export const cifrarJson = async (clave, obj, iv, ahora) => b64.a(await cifrarBytes(clave, new TextEncoder().encode(JSON.stringify(obj)), iv, ahora))
export const descifrarJson = async (clave, texto, ahora) => JSON.parse(new TextDecoder().decode(await descifrarBytes(clave, b64.de(texto), ahora)))

File: receptor/sincro/src/cifrado.rs
// receptor/sincro/src/cifrado.rs
//! AES-256-GCM con el mismo formato que `app/src/lib/cifrado.js`: IV (12) + cifrado con etiqueta;
//! el contenido lleva delante la hora en ms (8 bytes, big-endian). Datos adicionales fijos.
use aes_gcm::aead::{Aead, KeyInit, Payload};
use aes_gcm::{Aes256Gcm, Nonce};
use base64::{engine::general_purpose::STANDARD, Engine};
use std::time::{SystemTime, UNIX_EPOCH};

pub const AAD: &[u8] = b"canvas-sincro-v1";
pub const EDAD_MAX_MS: u64 = 5 * 60 * 1000;

#[derive(Debug, PartialEq, Eq)]
pub enum ErrorCifrado {
    /// La clave no es base64 de 32 bytes.
    Clave,
    /// Otra clave, o el mensaje fue modificado / está incompleto.
    Alterado,
    /// La hora del mensaje difiere más de EDAD_MAX_MS.
    Vencido,
}

pub fn ahora_ms() -> u64 {
    SystemTime::now().duration_since(UNIX_EPOCH).map(|d| d.as_millis() as u64).unwrap_or(0)
}

pub struct Clave(Aes256Gcm);

impl Clave {
    pub fn desde_base64(s: &str) -> Result<Self, ErrorCifrado> {
        let raw = STANDARD.decode(s.trim()).map_err(|_| ErrorCifrado::Clave)?;
        if raw.len() != 32 {
            return Err(ErrorCifrado::Clave);
        }
        Aes256Gcm::new_from_slice(&raw).map(Clave).map_err(|_| ErrorCifrado::Clave)
    }

    /// Clave nueva al azar y su texto base64 (para el QR de vinculación).
    pub fn nueva() -> (Self, String) {
        let mut k = [0u8; 32];
        getrandom::getrandom(&mut k).expect("sin fuente de azar");
        let s = STANDARD.encode(k);
        (Self::desde_base64(&s).expect("clave recién creada"), s)
    }

    pub fn cifrar_con(&self, bytes: &[u8], iv: [u8; 12], ahora: u64) -> Vec<u8> {
        let mut plano = ahora.to_be_bytes().to_vec();
        plano.extend_from_slice(bytes);
        let ct = self.0.encrypt(Nonce::from_slice(&iv), Payload { msg: &plano, aad: AAD }).expect("cifrar");
        let mut out = iv.to_vec();
        out.extend(ct);
        out
    }

    pub fn cifrar(&self, bytes: &[u8]) -> Vec<u8> {
        let mut iv = [0u8; 12];
        getrandom::getrandom(&mut iv).expect("sin fuente de azar");
        self.cifrar_con(bytes, iv, ahora_ms())
    }

    pub fn descifrar_en(&self, sobre: &[u8], ahora: u64) -> Result<Vec<u8>, ErrorCifrado> {
        if sobre.len() < 12 + 8 + 16 {
            return Err(ErrorCifrado::Alterado);
        }
        let plano = self
            .0
            .decrypt(Nonce::from_slice(&sobre[..12]), Payload { msg: &sobre[12..], aad: AAD })
            .map_err(|_| ErrorCifrado::Alterado)?;
        let t = u64::from_be_bytes(plano[..8].try_into().expect("8 bytes"));
        if ahora.abs_diff(t) > EDAD_MAX_MS {
            return Err(ErrorCifrado::Vencido);
        }
        Ok(plano[8..].to_vec())
    }

    pub fn descifrar(&self, sobre: &[u8]) -> Result<Vec<u8>, ErrorCifrado> {
        self.descifrar_en(sobre, ahora_ms())
    }

    pub fn cifrar_json(&self, v: &serde_json::Value) -> String {
        STANDARD.encode(self.cifrar(&serde_json::to_vec(v).expect("json")))
    }

    pub fn descifrar_json(&self, texto: &str) -> Result<serde_json::Value, ErrorCifrado> {
        let b = STANDARD.decode(texto.trim()).map_err(|_| ErrorCifrado::Alterado)?;
        serde_json::from_slice(&self.descifrar(&b)?).map_err(|_| ErrorCifrado::Alterado)
    }
}
